### PR TITLE
Add final layout to output quantum circuit from transpiler

### LINKED ISSUE #5350: Extracting permutation of classical registers from Quantum Circuit
Hi,

during routing/mapping passes input quantum registers could be permuted, the permutation logical -> physical qregs is saved in `quantum_circuit._layout` property and could be easily accessed by a user.

Similarly, classical cregs also become permuted after routing/mapping: physical cregs->logical cregs. This permutation can be extracted when applying measure gates `quantum_circuit.measure_all()`. E.g. when two cregs (0 and 1) are permuted than 
the measure gates will look like: 
`measure qreg[0] creg[1];`
`measure qreg[1] creg[0]`.

However, _when measure gates are not applied_, there is still could be implicit permutation of cregs after mapping/routing pass. It seems that there is no simple way to access information about cregs permutation, which could be inconvenient in some cases. For example this information is necessary to check the equivalence of original and transpiled circuits.
Is it possible to add this information as a property of QuantumCircuit object? Or there is some other way to access this information?

Thanks a lot!

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit expands the transpiler to return the final layout
permutation caused by the swap mapping in the routing stage. The routing
passes already return the final layout permutation caused by the swap
mapping to the property set (although for some passes it was only
setting it sometime, which this fixes). This commit just sets the
_final_layout attribute of QuantumCircuit objects to store this Layout
object. This lets user track the qubit permutation caused by
transpilation if they need.

Fixes #5350

### Details and comments

TODO:
- [x] Verify operator change actually works (I think I did the input mapping correctly but not sure the output transform being applied is correct)
- [x] Add tests
- [ ] Rewrite `Operator.from_circuit` to work natively on arrays instead of the super slow permutation matrix using `Permutation` class
- [x] Add release notes

### CODE REVIEW

**Comment by alexanderivrii** [comment]
Per @jakelishman's suggestion, I am posting here a code snippet that shows how `Permutations` are implicitly applied after routing, with the code snippet checking that the pre-routed and post-routed circuits are equivalent, 

```
qc = QuantumCircuit(3)
qc.cx(0, 1)
qc.cx(0, 2)
op1 = Operator(qc)

coupling_map = CouplingMap([[0, 1], [1, 0], [1, 2], [2, 1]])
sabre_swap = SabreSwap(coupling_map=coupling_map, heuristic="decay", seed=0)
pm = PassManager(sabre_swap)
tqc = pm.run(qc)
final_layout = sabre_swap.property_set['final_layout']

nq = qc.num_qubits
perm_pattern = [final_layout._v2p[v] for v in qc.qubits]

tqc_with_perm = tqc.copy()
tqc_with_perm.append(Permutation(nq, perm_pattern), range(nq))
op2 = Operator(tqc_with_perm)

print(op1 == op2)
```

**Comment by nonhermitian** [comment]
I do not think this holds when there is an initial layout that is non-trivial.

**Comment by jakelishman** [comment]
Paul: at worst, that code block is just missing an extra indirection through the `initial_layout`. The reason I'd suggested posting it here was just because Matthew's original attempt used `qargs` (which is a two-sided permutation), and Sasha had an example that came up elsewhere where he'd needed to apply this same one-sided permutation in more limited circumstances, and he'd written this.  It was just so Matthew could see (once he's back) approximately what we'd meant, and he could build it into the form he needed for this PR.

**Comment by alexanderivrii** [comment]
@nonhermitian, agreed, that code snippet is not meant to be general or to work when a target device has more qubits than the original circuit; the purpose is to emphasize @jakelishman's point that the output of a routing pass is a single-sided permutation and to give an example of how this permutation could be applied (sorry if this was not clear).

**Comment by nonhermitian** [comment]
Ahh ok, my confusion then.  Wholeheartedly agree that more people should understand routing as a permutation.

**Comment by mtreinish** [comment]
I separated out the fix for the initial layout permutation in #8802 we should probably rebase this on top of that since it changes the first half of the code in `from_circuit`.

File: test/python/quantum_info/operators/test_operator.py [hunk 1]
```diff
@@ -743,6 +743,78 @@ def test_from_circuit_constructor_reverse_embedded_layout(self):
...
+    def test_from_circuit_constructor_reverse_embedded_layout_from_transpile_with_registers(self):
+        """Test initialization from a circuit with an embedded final layout."""
+        # Test tensor product of 1-qubit gates
+        qr = QuantumRegister(3, name="test_reg")
+        circuit = QuantumCircuit(qr)
+        circuit.h(0)
+        circuit.x(1)
+        circuit.ry(np.pi / 2, 2)
+        output = transpile(circuit, initial_layout=[2, 1, 0])
+        print(output)
```
Comment (alexanderivrii) [inline test/python/quantum_info/operators/test_operator.py:769]: debug print

**Review by alexanderivrii — COMMENTED** [review]
Overall, this looks good to me. 

I may have mentioned this already somewhere, but there is still one very simple testcase where equivalence fails:
```
def test8():
    qc = QuantumCircuit(6)
    qc.swap(0, 1)
    qct = transpile(qc, optimization_level=3)
    op1 = Operator.from_circuit(qc)
    op2 = Operator.from_circuit(qct)
    equiv = matrix_equal(op1.data, op2.data, ignore_phase=True)
    assert(equiv)
```
Here the problem is due to `OptimizeSwapBeforeMeasure` transpiler pass throwing out the SWAP gate without updating the final_layout. 

I am planning to run a few more tests on randomly generated circuits and see what happens.

**Comment by alexanderivrii** [comment]
Here is one more random example where equivalence fails (and I haven't figured why yet): 
```
original_circuit = random_circuit(3, 2)
op1 = Operator.from_circuit(original_circuit)
coupling_map = CouplingMap([[1, 0], [0, 1], [0, 2], [2, 0]])
qct = transpile(original_circuit, coupling_map=coupling_map, optimization_level=3, seed_transpiler=1)
op2 = Operator.from_circuit(qct)
equiv = matrix_equal(op1.data, op2.data, ignore_phase=True)
assert(equiv)
```
BTW, if you think it makes sense, I am not against merging this PR as is, and handling the remaining equivalence checking issues in a follow-up PR.

File: test/python/quantum_info/operators/test_operator.py (hunk 1 above)
Comment (mtreinish) [reply test/python/quantum_info/operators/test_operator.py:769]:
```suggestion
```

File: qiskit/quantum_info/operators/operator.py [hunk 2]
```diff
@@ -224,8 +224,14 @@ def from_circuit(cls, circuit, ignore_set_layout=False, layout=None):
...
                 the virtual bits in the :class:`~.Layout` must be present in the
                 ``circuit`` input.
+            final_layout (Layout): If specified this kwarg can be used to represent the
+                output permutation caused by swap insertions during the routing stage
+                of the transpiler.
         Returns:
             Operator: An operator representing the input circuit
+
+        Raises:
+            KeyError: If a circuit has an empty embedded.
```
Comment (pedrorrivero) [inline qiskit/quantum_info/operators/operator.py:234]: Not sure if this phrasing is correct, seems to be missing a word.

File: qiskit/transpiler/layout.py [hunk 3]
```diff
@@ -374,3 +374,4 @@ class TranspileLayout:
 
     initial_layout: Layout
     input_qubit_mapping: Dict[Qubit, int]
+    final_layout: Optional[Layout] = None
```
Comment (pedrorrivero) [inline qiskit/transpiler/layout.py:377]: What does it mean for `final_layout` to be optional? Is it that transpiler passes could swap qubits and not report it, or that they can leave this unset if they are not performing any swaps?

**Review by pedrorrivero — COMMENTED** [review]
Thanks for putting this together, greatly useful!

File: qiskit/transpiler/layout.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/layout.py:377]: It primarily means the latter, if `final_layout` is None then there wasn't a permutation being caused by the routing stage in the compiler. Basically, the routing pass was skipped. There is also a backwards compatibility component to making this optional as well, but that's less likely to come up in practice outside of the routing pass not run.

File: qiskit/quantum_info/operators/operator.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/quantum_info/operators/operator.py:234]: heh, good catch. Yeah it's missing the word layout, however this actually isn't needed anymore. I think it's leftover from an earlier revision. I'll just remove this.

**Review by ikkoham — COMMENTED** [review]
Overall LGTM!

File: qiskit/quantum_info/operators/operator.py [hunk 4]
```diff
@@ -252,6 +262,16 @@ def from_circuit(cls, circuit, ignore_set_layout=False, layout=None):
         # Convert circuit to an instruction
         instruction = circuit.to_instruction()
         op._append_instruction(instruction, qargs=qargs)
+        # If final layout is set permute output indices based on layout
+        if final_layout is not None:
+            # TODO: Do this without the intermediate Permutation object by just
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/quantum_info/operators/operator.py:267]: I think we can save it for a follow up, unless you have an easy way to do it now. I don't have the bandwidth to investigate this further anymore.

File: qiskit/quantum_info/operators/operator.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/quantum_info/operators/operator.py]: Removed in: https://github.com/Qiskit/qiskit-terra/pull/8597/commits/a0c5521a717b1bb84d9422173f2718e3c004f258

**Review by alexanderivrii — APPROVED** [review]
I was happy with this PR when I have reviewed it in October, and since then it did not change much except for a few minor fixes. So I think this is good to go. LGTM. 

Regarding the "TODO" comment. If #9157 gets merged, one should be able to immediately replace the more expensive `perm_op = Operator(Permutation(len(final_physical_to_virtual), perm_pattern))` (which first synthesizes a quantum circuit for the permutation and then builds an operator for this circuit) by the less expensive `perm_op = Operator(PermutationGate(perm_pattern))` (which builds an operator from the permutation pattern directly). Actually, the `PermutationGate`'s `__array__` method works out which of the entries in the permutation's unitary matrix are equal to 1, and it should be straightforward to adapt it to replace the line `op &= perm_op` by direct swaps of the rows (or maybe the columns) of the `op` matrix. If you wish I can quickly push a function `apply_permutation` to `Operator` class. But it can also wait till after the release.

**Review by kdk — APPROVED** [review]

**Pull request merged**